File: python/src/day01.py
#!/usr/bin/env python

from pathlib import Path
from typing import List, Optional

INPUT_FILE = Path.cwd().parent / "inputs" / "day01.txt"


def transform_input(input: str) -> List[int]:
    return [int(x) for x in input.split()]


def part1(numbers: List[int], target=2020) -> Optional[int]:
    complements = set()

    for x in numbers:
        if x in complements:
            return x * (target - x)

        complements.add(target - x)

    return None


def part2(numbers: List[int]) -> int:
    for i, x in enumerate(numbers):
        if yz := part1(numbers[i + 0:], target=2020 - x):
            return x * yz

    raise Exception("No triplet found.")


if __name__ == "__main__":
    with INPUT_FILE.open("r") as f:
        input = transform_input(f.read())

    print(f"Part 1: {part1(input)}")
    print(f"Part 2: {part2(input)}")

File: rust/src/day01.rs
use std::collections::HashSet;

static INPUT: &str = include_str!("../../inputs/day01.txt");

pub fn run() {
    let numbers = transform_input(INPUT);

    println!("Part 1: {}", part1(&numbers));
    println!("Part 2: {}", part2(&numbers));
}

fn transform_input(input: &str) -> Vec<i32> {
    input.lines().filter_map(|x| x.parse().ok()).collect()
}

fn part1(numbers: &[i32]) -> i32 {
    find_two_sum(numbers, 2020).unwrap()
}

fn find_two_sum(numbers: &[i32], target: i32) -> Option<i32> {
    let mut complements: HashSet<i32> = HashSet::new();

    for x in numbers.iter() {
        if complements.contains(x) {
            return Some(x * (target - x));
        }

        complements.insert(target - x);
    }

    None
}

fn part2(numbers: &[i32]) -> i32 {
    for x in numbers.iter() {
        if let Some(yz) = find_two_sum(numbers, 2020 - x) {
            return x * yz;
        }
    }

    panic!("No triplet found.");
}
